### PR TITLE
Add mobile: endpoint for installing custom certificate

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
This still has to be tested properly, but I just wanted to demonstrate the idea first.

### CODE REVIEW

File: lib/commands/certificate.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  } catch (err) {
+    throw new Error(`Cannot parse the certificate. Is it valid and base64-encoded? ` +
+                    `Original error: ${err.message}`);
+  }
+  const displayName = _.replace(_.last(altNames), /\*/g, '_');
+  if (_.isEmpty(displayName)) {
+    throw new Error(`Cannot parse alternate names from the certificate. ` +
+                    `Original value: ${JSON.stringify(altNames)}`);
+  }
+  const uuid = () => UUID.create().hex.toUpperCase();
```
Comment (jlipps) [inline lib/commands/certificate.js]: even though it's a different case, i'm not sure i like `uuid` having the same name as `UUID`?

Comment (jlipps) [reply lib/commands/certificate.js]: maybe `getUUID()`?

File: lib/commands/certificate.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+async function installCertificateInPreferences (driver) {
+  // Accept Safari alert
+  await clickElement(driver, {
+    type: 'accessibility id',
+    value: 'Allow'
+  });
+
+  await B.delay(2000);
```
Comment (jlipps) [inline lib/commands/certificate.js]: shouldn't the `waitForCondition` in `clickElement` make it unnecessary to have a static wait here?

File: lib/commands/certificate.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+/**
+ * @typedef {Object} CertificateInstallationOptions
+ *
+ * @property {!string} content - Base64-encoded content of the public certificate
+ */
+
+/**
+ * Installs a custom certificate onto the device.
```
Comment (jlipps) [inline lib/commands/certificate.js]: you do realize this is insane :-)

File: lib/commands/certificate.js [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+ *
+ * @param {CertificateInstallationOptions} opts
+ */
+commands.mobileInstallCertificate = async function (opts = {}) {
+  const {content} = opts;
+  if (_.isEmpty(content)) {
+    throw new Error('Certificate content should not be empty');
+  }
+
+  const staticContentRoot = path.resolve(
```
Comment (jlipps) [inline lib/commands/certificate.js]: if we make this a real tested feature, we could probably expose the static root dir as part of the package so we don't need to go hunting for it

File: lib/commands/certificate.js (hunk 1 above)
Comment (mykola-mokhnach) [reply lib/commands/certificate.js]: ok, fixed

**Comment by mykola-mokhnach** [comment]
I've tested the implementation and it worked. So, the only thing left is to incorporate static dir constant from the base driver as soon as it is published

**Comment by mykola-mokhnach** [comment]
@KazuCocoa FYI

File: lib/commands/certificate.js [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    PayloadType: 'Configuration',
+    PayloadUUID: getUUID(),
+    PayloadVersion: 1
+  };
+}
+
+async function clickElement (driver, locator, timeout = 5000) {
+  let element = null;
+  try {
+    await waitForCondition(async () => {
```
Comment (imurchie) [inline lib/commands/certificate.js]: `retryInterval` was made for these sorts of situations, where an exception is thrown:
```js
try {
  element = await retryInterval(timeout / 500, 500, async () => {
    return await driver.findNativeElementOrElements(locator.type, locator.value, false);
  });
} catch (err) {
  throw new Error(`Cannot find ${JSON.stringify(locator)} within ${timeout}ms timeout`);
}
```

File: lib/commands/certificate.js [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        await this.activateApp(this.opts.bundleId);
+      } catch (e) {
+        log.warn(`Cannot restore the application '${this.opts.bundleId}'. ` +
+                 `Original error: ${e.message}`);
+      }
+    }
+
+    return (await fs.readFile(configPath)).toString('base64');
+  } finally {
+    if (await fs.exists(configPath)) {
```
Comment (imurchie) [inline lib/commands/certificate.js]: No need to check. `await fs.rimraf(configPath);` will either delete the directory, if it exists, or quietly go on its way without doing anything, if it does not exist.

File: lib/commands/certificate.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+  }
+  const displayName = _.replace(subject.commonName, /\*/g, '_');
+  if (_.isEmpty(displayName)) {
+    throw new Error(`Cannot parse commonName field from the certificate subject. ` +
+                    `Original value: ${JSON.stringify(subject)}`);
+  }
+  const getUUID = () => UUID.create().hex.toUpperCase();
+  const contentUuid = getUUID();
+  return {
+    PayloadContent: [{
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply lib/commands/certificate.js]: yes. I was not able to find any normal documentation on this format. There is `Apple Configurator 2`  in the AppStore, which makes such files.

[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply lib/commands/certificate.js]: I've added a link to https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/iPhoneOTAConfiguration/Introduction/Introduction.html
in the documentation.

This is, at least, something

[the remark this answers is not in the document]
Comment (imurchie) [reply lib/commands/certificate.js]: Sounds good.

**Comment by KazuCocoa** [comment]
interesting 🆒

File: lib/commands/certificate.js [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+  return true;
+}
+
+async function installCertificateInPreferences (driver) {
+  // Accept Safari alert
+  await clickElement(driver, {
+    type: 'accessibility id',
+    value: 'Allow'
+  });
+  await B.delay(2000);
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply lib/commands/certificate.js]: Added explanation comments

**Comment by pr4bh4sh** [comment]
Definitely off topic comment but can't resist.

I suggested this approach to someone and I can still remember the death stare I received. 
To that someone, here it is "[I told you so](https://www.youtube.com/embed/NzpR4eRxd8E?start=111&end=118)".

PS. No offence to anyone in this thread.

**Comment by mykola-mokhnach** [comment]
@pr4bh4sh Ideas and implementations are slightly different things. It would be ideal if you could create your own PR with the actual implementation. We are always happy to accept contributions. Also, if some feature is really important for one, then it is always possible to hire a person, who could do the job.

**Pull request merged**